File: libs/graphicsgeo/src/main/kotlin/mono/graphics.geo/MousePointer.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.graphics.geo

/**
 * A sealed class which indicates mouse event pointer types.
 */
sealed interface MousePointer {
    val boardCoordinate: Point

    object Idle : MousePointer {
        override val boardCoordinate: Point
            get() = Point.ZERO
    }

    data class Move(
        override val boardCoordinate: Point,
        val pointPx: Point
    ) : MousePointer

    data class Down(
        override val boardCoordinate: Point,
        val pointPx: Point,
        val isWithShiftKey: Boolean
    ) : MousePointer

    data class Drag(
        val mouseDownPoint: Point,
        override val boardCoordinate: Point,
        val boardCoordinateF: PointF,
        val isWithShiftKey: Boolean
    ) : MousePointer

    data class Up(
        val mouseDownPoint: Point,
        override val boardCoordinate: Point,
        val boardCoordinateF: PointF,
        val isWithShiftKey: Boolean
    ) : MousePointer

    data class Click(
        override val boardCoordinate: Point,
        val isWithShiftKey: Boolean
    ) : MousePointer

    data class DoubleClick(override val boardCoordinate: Point) : MousePointer
}

File: monosketch-svelte/src/lib/mono/workspace/mouse/mouse-pointer.ts
import { Point, PointF } from '$libs/graphics-geo/point';

export enum MousePointerType {
    IDLE,
    MOVE,
    DOWN,
    UP,
    DRAG,
    CLICK,
    DOUBLE_CLICK,
}

export type MousePointerIdle = { type: MousePointerType.IDLE };
export type MousePointerMove = { type: MousePointerType.MOVE; boardCoordinate: Point; clientCoordinate: Point };
export type MousePointerDown = {
    type: MousePointerType.DOWN;
    boardCoordinate: Point;
    clientCoordinate: Point;
    isWithShiftKey: boolean
};
export type MousePointerDrag = {
    type: MousePointerType.DRAG;
    boardCoordinate: Point;
    boardCoordinateF: PointF;
    mouseDownBoardCoordinate: Point;
    isWithShiftKey: boolean
};
export type MousePointerUp = {
    type: MousePointerType.UP;
    boardCoordinate: Point;
    boardCoordinateF: PointF;
    mouseDownBoardCoordinate: Point;
    isWithShiftKey: boolean
};
export type MousePointerClick = { type: MousePointerType.CLICK; boardCoordinate: Point; isWithShiftKey: boolean };
export type MousePointerDoubleClick = { type: MousePointerType.DOUBLE_CLICK; boardCoordinate: Point };


/**
 * Type representing different mouse event pointer states
 */
export type MousePointer =
    | MousePointerIdle
    | MousePointerMove
    | MousePointerDown
    | MousePointerUp
    | MousePointerDrag
    | MousePointerClick
    | MousePointerDoubleClick;


export namespace MousePointer {
    export const idle: MousePointerIdle = { type: MousePointerType.IDLE };

    export function move(boardCoordinate: Point, clientCoordinate: Point): MousePointerMove {
        return { type: MousePointerType.MOVE, boardCoordinate, clientCoordinate };
    }

    export function down(boardCoordinate: Point, clientCoordinate: Point, isWithShiftKey: boolean): MousePointerDown {
        return { type: MousePointerType.DOWN, boardCoordinate, clientCoordinate, isWithShiftKey };
    }

    export function up(
        boardCoordinate: Point,
        boardCoordinateF: PointF,
        mouseDownBoardCoordinate: Point,
        isWithShiftKey: boolean,
    ): MousePointerUp {
        return {
            type: MousePointerType.UP,
            boardCoordinate,
            boardCoordinateF,
            mouseDownBoardCoordinate,
            isWithShiftKey,
        };
    }

    export function drag(
        boardCoordinate: Point,
        boardCoordinateF: PointF,
        mouseDownBoardCoordinate: Point,
        isWithShiftKey: boolean,
    ): MousePointerDrag {
        return {
            type: MousePointerType.DRAG,
            boardCoordinate,
            boardCoordinateF,
            mouseDownBoardCoordinate,
            isWithShiftKey,
        };
    }

    export function click(boardCoordinate: Point, isWithShiftKey: boolean): MousePointerClick {
        return { type: MousePointerType.CLICK, boardCoordinate, isWithShiftKey };
    }

    export function doubleClick(boardCoordinate: Point): MousePointerDoubleClick {
        return { type: MousePointerType.DOUBLE_CLICK, boardCoordinate };
    }
}
